File: Lintcode/60. Search Insert Position/LeetCode_35.py
class LeetCode_35(object):
    def searchInsert(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: int
        """
        length = len(nums)
        if len == 0 or nums[0] >= target:
            return 0
        elif nums[length - 1] < target:
            return length;
        
        left = 0
        right = length - 1
        
        while left < right:
            mid = (left + right) // 2
            if nums[mid] < target:
                left = mid + 1
            else:
                right = mid
        return left
        
        

File: Lintcode/60. Search Insert Position/LeetCode_35.java
class LeetCode_35 {
    public int searchInsert(int[] nums, int target) {
        if (nums == null || nums.length == 0) {
            return -1;
        }
        //if (nums[nums.length - 1] < target) {
        //    return nums.length;
        // }
        int left = 0, right = nums.length - 1;
        while (left <= right) {
            int mid = left + (right - left) / 2; // mid belongs to [left, mid] 
            if (nums[mid] == target) {
                return mid;
            } else if (nums[mid] < target) {
                left = mid + 1;
            } else {
                right = mid - 1;
            }
        }
        return left; //note left could be the largest
    }
}